test/integration/gateway_test.go

```diff
@@ -38,74 +37,34 @@ const (
 )
 
 func TestUnmanagedGatewayBasics(t *testing.T) {
-	ns, cleanup := namespace(t)
-	defer cleanup()
+	var gw *gatewayv1alpha2.Gateway
 
-	t.Log("generating a gateway kubernetes client and gathering test data")
+	ns, cleaner := setup(t)
+	defer func() { assert.NoError(t, cleaner.Cleanup(ctx)) }()
+
+	t.Log("gathering test data and generating a gateway kubernetes client")
 	pubsvc, err := env.Cluster().Client().CoreV1().Services(controllerNamespace).Get(ctx, "ingress-controller-kong-proxy", metav1.GetOptions{})
 	require.NoError(t, err)
-	c, err := gatewayclient.NewForConfig(env.Cluster().Config())
-	require.NoError(t, err)
-
-	t.Log("deploying a supported gatewayclass to the test cluster")
-	supportedGatewayClass := &gatewayv1alpha2.GatewayClass{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: uuid.NewString(),
-		},
-		Spec: gatewayv1alpha2.GatewayClassSpec{
-			ControllerName: gateway.ControllerName,
-		},
-	}
-	supportedGatewayClass, err = c.GatewayV1alpha2().GatewayClasses().Create(ctx, supportedGatewayClass, metav1.CreateOptions{})
+	gatewayClient, err := gatewayclient.NewForConfig(env.Cluster().Config())
 	require.NoError(t, err)
 
-	defer func() {
-		t.Log("cleaning up gatewayclasses")
-		assert.NoError(t, c.GatewayV1alpha2().GatewayClasses().Delete(ctx, supportedGatewayClass.Name, metav1.DeleteOptions{}))
-	}()
-
-	t.Log("deploying a gateway to the test cluster using unmanaged gateway mode")
-	gw := &gatewayv1alpha2.Gateway{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "kong",
-			Annotations: map[string]string{
-				unmanagedAnnotation: "true", // trigger the unmanaged gateway mode
-			},
-		},
-		Spec: gatewayv1alpha2.GatewaySpec{
-			GatewayClassName: gatewayv1alpha2.ObjectName(supportedGatewayClass.Name),
-			Listeners: []gatewayv1alpha2.Listener{{
-				Name:     "http",
-				Protocol: gatewayv1alpha2.HTTPProtocolType,
-				Port:     gatewayv1alpha2.PortNumber(80),
-			}},
-		},
-	}
-	gw, err = c.GatewayV1alpha2().Gateways(ns.Name).Create(ctx, gw, metav1.CreateOptions{})
-	require.NoError(t, err)
-
-	t.Log("verifying that the Gateway object gets scheduled by the controller")
-	require.Eventually(t, func() bool {
-		gw, err = c.GatewayV1alpha2().Gateways(ns.Name).Get(ctx, gw.Name, metav1.GetOptions{})
-		require.NoError(t, err)
-		for _, cond := range gw.Status.Conditions {
-			if cond.Reason == string(gatewayv1alpha2.GatewayReasonScheduled) {
-				return true
-			}
-		}
-		return false
-	}, gatewayUpdateWaitTime, time.Second)
+	t.Log("deploying a new gateway")
+	gateway, err := DeployGateway(ctx, gatewayClient, ns.Name, managedGatewayClassName)
+	exitOnErr(err)
```

Comment:
I needed to check this 😅 but `os.Exit()` (which is called in `exitOnErr`) prevents deferred functions and `t.Cleanup()` functions from running which is problematic because we want to clean up the created k8s resources.

```go
package main

import (
	"os"
	"testing"
)

func TestX(t *testing.T) {
	t.Cleanup(func() {
		t.Log("cleanup")
		if t.Failed() {
			t.Log("cleanup if")
		}
	})

	defer func() {
		t.Log("defer")
		if t.Failed() {
			t.Log("defer if")
		}
	}()

	os.Exit(1)
}
```

Fun fact: `os.Exit(0)` will panic in the test 🙃 

```
=== RUN   TestX
    /Users/patryk.malek@konghq.com/code_/own/testfail/x_test.go:17: defer
    /Users/patryk.malek@konghq.com/code_/own/testfail/x_test.go:10: cleanup
--- FAIL: TestX (0.00s)
panic: unexpected call to os.Exit(0) during test [recovered]
	panic: unexpected call to os.Exit(0) during test

goroutine 4 [running]:
testing.tRunner.func1.2({0x1010c0ea0, 0x1010e1980})
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/testing/testing.go:1389 +0x1c8
testing.tRunner.func1()
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/testing/testing.go:1392 +0x384
panic({0x1010c0ea0, 0x1010e1980})
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/runtime/panic.go:838 +0x204
os.Exit(0x0)
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/os/proc.go:68 +0x58
testfail.TestX(0x140001381a0)
	/Users/patryk.malek@konghq.com/code_/own/testfail/x_test.go:23 +0xac
testing.tRunner(0x140001381a0, 0x1010e1818)
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/testing/testing.go:1439 +0x110
created by testing.(*T).Run
	/Users/patryk.malek@konghq.com/.gvm/gos/go1.18.3/src/testing/testing.go:1486 +0x300
FAIL	testfail	0.006s
```

Hence I suggest to use `require.NoError(t, err)`

```suggestion
	require.NoError(t, err)
```